File: rr/src/glicko.rs
use uuid::Uuid;
use serde::{Serialize,Deserialize};
use serde_json;

use chrono::DateTime;
use chrono::offset::Utc;

/*
 * https://en.wikipedia.org/wiki/Glicko_rating_system
 */

/****************************************************************
 *							  CONSTANTS                         *
 ****************************************************************/

const INIT_RATING: f32 = 1500.0;
const INIT_RD: f32 = 350.0;

// Rating periods to max uncertainty
const MU: f32 = 100.0; // TODO: This is a guess

// Average player's ratings deviation
const AVG_RD: f32 = 50.0; // TODO: This is a guess

// ln(10)/400
const Q: f32 = 0.00575646273;
const PI: f32 = 3.14159;

fn get_diff_in_weeks(d0: DateTime<Utc>, d1: DateTime<Utc>) -> u64 {
    (d1.timestamp() - d0.timestamp()) as u64
}

fn g(rd_i: f32) -> f32 {
    1.0 / (1.0 + (3.0*Q.powf(2.0)*rd_i.powf(2.0))/PI.powf(2.0)).sqrt()
}

fn e(r_0: f32, r_i: f32, rd_i:f32) -> f32 {
    1.0 / (1.0 + 10.0_f32.powf((g(rd_i)*(r_0 - r_i)) / 400.0))
}

fn calc_d2(r0: f32, games: &Vec<Game>) -> f32 {
    1.0 / (Q.powf(2.0) * games.iter()
           .fold(0.0, |sum: f32, o| sum + g(o.rd).powf(2.0)
                 * e(r0, o.r, o.rd)
                 * (1.0 - e(r0, o.r, o.rd))))
}

fn rd_add_uncertainty(rd_0: f32, t: u64) -> f32 {
    let c : f32 = ((MU.powf(2.0) - AVG_RD.powf(2.0)) / 100.0).sqrt();
    (rd_0.powf(2.0) + (t as f32)*c.powf(2.0)).sqrt()
}

fn update_rating(r0: f32, rd: f32, d2: f32, games: &Vec<Game>) -> f32 {
    r0 + (Q/(1.0/rd.powf(2.0) + 1.0/d2))*games.iter()
        .fold(0.0, |sum, o| sum + g(o.rd)*(o.s - e(r0, o.r, o.rd)))
}

fn update_rd(rd_0: f32, d2: f32) -> f32 {
    (1.0 / (1.0 / rd_0.powf(2.0) + 1.0 / d2)).sqrt()
}

#[derive(Serialize, Deserialize)]
struct Game {
    s: f32,
    r: f32,
    rd: f32
}

#[derive(Serialize, Deserialize)]
pub struct Player {
    pub name: String,
    pub id: Uuid,
    pub r: f32,
    rd: f32,
    last_date: DateTime<Utc>,
    games: Vec<Game>
}

impl Player {
    pub fn new(name: String) -> Player {
        Player {
            name,
            id: Uuid::new_v4(),
            r: INIT_RATING,
            rd: INIT_RD,
            last_date: Utc::now(),
            games: Vec::new()
        }
    }

    pub fn update_rating(&mut self) {
        let t = get_diff_in_weeks(self.last_date, Utc::now());
        let rd0 = rd_add_uncertainty(self.rd, t);
        let d2 = calc_d2(rd0, &self.games);

        self.r = update_rating(self.r, rd0, d2, &self.games);
        self.rd = update_rd(rd0, d2);

        self.last_date = Utc::now();
        self.games.clear();
    }

    pub fn report_win(&mut self, opponent: &Player) {
        self.games.push(Game { s: 1.0, r: opponent.r, rd: opponent.rd });
    }

    pub fn report_loss(&mut self, opponent: &Player) {
        self.games.push(Game { s: 0.0, r: opponent.r, rd: opponent.rd });
    }

    pub fn report_draw(&mut self, opponent: &Player) {
        self.games.push(Game { s: 0.5, r: opponent.r, rd: opponent.rd });
    }

    pub fn from_json(s: &str) -> Result<Player, serde_json::Error> {
        serde_json::from_str(s)
    }
}

File: glicko.js
import crypto from 'crypto';

/*
 * https://en.wikipedia.org/wiki/Glicko_rating_system
 */

/********************************************************************
 *							  CONSTANTS                             *
 ********************************************************************/

// Rating periods to max uncertainty
const mu = 100; // TODO: This is a guess
// Average player's ratings deviation
const avg_rd = 50; // TODO: This is a guess

const c = Math.sqrt((mu**2 - avg_rd**2)/100);
// ln(10)/400
const q = 0.00575646273;

/********************************************************************
 *						  HELPER FUNCTIONS		                    *
 ********************************************************************/

const get_diff_in_weeks = (d0, d1) => {
	// Get difference in milliseconds
	const diff = d1.getTime() - d0.getTime();

	// Convert to num days and return
	return Math.floor(diff / (1000 * 3600 * 24 * 7));
}

// g(RD_i)
const g = (RD_i) => 1/Math.sqrt(1 + (3 * q**2 * RD_i**2)/Math.PI**2);
// E(s|r_0, r_i, RD_i)
const E = (r_0, r_i, RD_i) => 1 / 
	(1 + 10**((g(RD_i)*(r_0 - r_i))/-400))

const calc_d2 = (r0, games) => 1/(q**2 * games.reduce((sum, opp) => 
	sum += g(opp.rd)**2 * E(r0, opp.r, opp.rd) * (1 - E(r0, opp.r, opp.rd)),
	0));

/*
 * Primary steps
 */

/** 
 * Step 1: Determine ratings deviation
 */
const RD_add_uncertainty = (rd0, t) => Math.min(
	Math.sqrt(rd0**2 + t*c**2),
	350
);

/**
 * Step 2: Determine new rating
 */
const update_rating = (r0, rd, d2, games) => 
	r0 + (q/(1/rd**2 + 1/d2))*games.reduce((sum, opp) => 
		sum += g(opp.rd)*(opp.s - E(r0, opp.r, opp.rd)),
		0);

/**
 * Step 3: Determine new ratings deviation
 */
const update_rd = (rd0, d2) => Math.sqrt(1/(1/rd0**2 + 1/d2));

/********************************************************************
 *							HELPER CLASS							*
 ********************************************************************/

class Game {
	constructor(s, r, rd) {
		this.s = s;
		this.r = r;
		this.rd = RD_add_uncertainty(rd, new Date());
	}
}

/********************************************************************
 *							 EXPORTS								*
 ********************************************************************/

export class Player {
	constructor(name, id = crypto.randomUUID(), r = 1500, rd = 350, last_date = new Date()) {
		this.name = name;
		this.id = id;
		this.r = r;
		this.rd = rd;
		this.last_date = last_date;
		this.games = [];
	}

	updateRating()  {
		const t = get_diff_in_weeks(this.last_date, new Date());
		const rd0 = RD_add_uncertainty(this.rd, t);
		const d2 = calc_d2(rd0, this.games);

		this.r = update_rating(this.r, rd0, d2, this.games);
		this.rd = update_rd(rd0, d2);
		this.games = [];
		this.last_date = new Date();

	}

	reportWin(opponent) {
		this.games.push(new Game(1, opponent.r, opponent.rd));
	}

	reportLoss(opponent) {
		this.games.push(new Game(0, opponent.r, opponent.rd));
	}
	
	reportDraw(opponent) {
		this.games.push(new Game(0.5, opponent.r, opponent.rd));
	}
}

Player.fromJSON = (json) => 
	new Player(json.name, json.id, json.r, json.rd, new Date(json.last_date));
